src/django_components/component.py

```diff
@@ -1362,28 +2260,20 @@ def _render_impl(
         # or inspect the Context object, your screen is filled with the deeply nested ComponentContext objects.
         component_context_cache[render_id] = component_ctx
 
-        # Allow to access component input and metadata like component ID from within these hook
+        # Allow to access component input and metadata like component ID from within data methods
         with self._with_metadata(metadata):
+            template_data, js_data, css_data = self._call_data_methods(args_list, kwargs_dict, slots_dict, context)
```

Comment:
Moved the calling of `get_template_data()`, `get_js_data()`, etc into `_call_data_methods()`.